Source/driver/Derive.cpp

```diff
@@ -396,418 +380,520 @@ extern "C"
-                  AMREX_INT_ANYD(domain_lo), AMREX_INT_ANYD(domain_hi),
-                  AMREX_REAL_ANYD(delta));
+        Real deninv = 1.0_rt/dat(i,j,k,0);
 
+        der(i,j,k,0) = std::sqrt( (dat(i,j,k,1) * dat(i,j,k,1) +
+                                   dat(i,j,k,2) * dat(i,j,k,2) +
+                                   dat(i,j,k,3) * dat(i,j,k,3)) ) * deninv;
+      });
     }
 
-    void ca_dermaggrav(Real* der, const int* der_lo, const int* der_hi, const int* nvar,
-                       const Real* data, const int* data_lo, const int* data_hi, const int* ncomp,
-                       const int* lo, const int* hi,
-                       const int* domain_lo, const int* domain_hi,
-                       const Real* delta, const Real* xlo,
-                       const Real* time, const Real* dt, const int* bcrec, 
-                       const int* level, const int* grid_no)
+
+    void ca_dermaggrav(const Box& bx, FArrayBox& derfab, int dcomp, int /*ncomp*/,
+                       const FArrayBox& datfab, const Geometry& geomdata,
+                       Real /*time*/, const int* /*bcrec*/, int /*level*/)
     {
 
-        IntVect ilo(D_DECL(lo[0], lo[1], lo[2]));
-        IntVect ihi(D_DECL(hi[0], hi[1], hi[2]));
+      auto const dat = datfab.array();
+      auto const der = derfab.array();
 
-        Box bx(ilo, ihi);
+      amrex::ParallelFor(bx,
+      [=] AMREX_GPU_DEVICE (int i, int j, int k) noexcept
+      {
 
-#pragma gpu box(bx)
-        dermaggrav(der, AMREX_INT_ANYD(der_lo), AMREX_INT_ANYD(der_hi), *nvar,
-                   data, AMREX_INT_ANYD(data_lo), AMREX_INT_ANYD(data_hi), *ncomp,
-                   AMREX_INT_ANYD(lo), AMREX_INT_ANYD(hi),
-                   AMREX_INT_ANYD(domain_lo), AMREX_INT_ANYD(domain_hi),
-                   AMREX_REAL_ANYD(delta));
+        der(i,j,k,0) = std::sqrt(dat(i,j,k,0)*dat(i,j,k,0) +
+                                 dat(i,j,k,1)*dat(i,j,k,1) +
+                                 dat(i,j,k,2)*dat(i,j,k,2));
 
+      });
     }
 
-    void ca_derradialvel(Real* der, const int* der_lo, const int* der_hi, const int* nvar,
-                         const Real* data, const int* data_lo, const int* data_hi, const int* ncomp,
-                         const int* lo, const int* hi,
-                         const int* domain_lo, const int* domain_hi,
-                         const Real* delta, const Real* xlo,
-                         const Real* time, const Real* dt, const int* bcrec, 
-                         const int* level, const int* grid_no)
+    void ca_derradialvel(const Box& bx, FArrayBox& derfab, int dcomp, int /*ncomp*/,
+                         const FArrayBox& datfab, const Geometry& geomdata,
+                         Real /*time*/, const int* /*bcrec*/, int /*level*/)
     {
 
-        IntVect ilo(D_DECL(lo[0], lo[1], lo[2]));
-        IntVect ihi(D_DECL(hi[0], hi[1], hi[2]));
+      auto const dat = datfab.array();
```

Comment:
For the derived fields that depend on location (radial velocity and angular momentum) we should use the `position` function rather than computing the position with `problo + dx * (idx + 0.5))`. The `position` function wraps around appropriately for periodic boundaries and this is relevant for correct evaluation of angular momentum for periodic problems.

Reply:
Actually, since this port is faithful to the Fortran, we can do that in a separate PR. I'm not sure if it even matters for the derive functions.